File: src/pty_host_client.py
"""PTY host client — webpty connects to the pty-host daemon over a Unix
socket. If the daemon isn't running we detach-spawn it (POSIX only).

Line-delimited JSON over a Unix socket; PTY output is base64 in the payload.
Uses asyncio so the HTTP/WS server can await host operations without
blocking its event loop.
"""
from __future__ import annotations

import asyncio
import base64
import json
import os
import socket
import sys

from logging_util import log_error

_HERE = os.path.dirname(os.path.abspath(__file__))
HOST_SCRIPT = os.path.join(_HERE, "pty_host.py")

PIPE_NAME = (
    os.environ.get("WEBPTY_PTY_HOST_PIPE")
    or ("/tmp/webpty-pty-host.sock" if os.name == "posix" else "webpty-pty-host")
)


class PtyHostError(Exception):
    pass


class PtyHostClient:
    def __init__(self) -> None:
        self.reader: asyncio.StreamReader | None = None
        self.writer: asyncio.StreamWriter | None = None
        self._req_id = 0
        self._pending: dict[int, asyncio.Future] = {}
        self.server_version: int | None = None
        self._connected = False
        self._connect_lock = asyncio.Lock()
        self._listeners: dict[str, list] = {"output": [], "exit": [], "disconnect": []}

    @property
    def connected(self) -> bool:
        """True while the socket to pty-host is open (see `_read_loop`)."""
        return self._connected

    # --- event listeners ---------------------------------------------------
    def on(self, event: str, cb) -> None:  # type: ignore[no-untyped-def]
        self._listeners.setdefault(event, []).append(cb)

    def off(self, event: str, cb) -> None:  # type: ignore[no-untyped-def]
        self._listeners.setdefault(event, []).remove(cb)

    def _emit(self, event: str, *args) -> None:  # type: ignore[no-untyped-def]
        for cb in list(self._listeners.get(event, [])):
            cb(*args)

    # --- connection ----------------------------------------------------------
    async def connect(self) -> None:
        if self._connected:
            return
        async with self._connect_lock:
            if self._connected:
                return
            await _ensure_host_running()
            reader, writer = await asyncio.open_unix_connection(PIPE_NAME)
            self.reader = reader
            self.writer = writer
            self._connected = True
            self._reader_task = asyncio.create_task(self._read_loop())

    async def _read_loop(self) -> None:
        assert self.reader is not None
        try:
            # Read without StreamReader's 64KB line limit: pty-host frames can
            # exceed it when many 32KB merged chunks flush in one line (base64
            # ≈ 44KB each), which crashed readuntil() with LimitOverrunError
            # and killed the whole host connection — sessions then went
            # unresponsive until the monitor reconnected (the reported
            # "反应不快速" slowness).
            buf = bytearray()
            while True:
                chunk = await self.reader.read(65536)
                if not chunk:
                    break
                buf += chunk
                while True:
                    nl = buf.find(b"\n")
                    if nl < 0:
                        break
                    line = bytes(buf[:nl])
                    del buf[:nl + 1]
                    line = line.decode("utf-8", "replace").strip()
                    if not line:
                        continue
                    try:
                        msg = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    self._on_message(msg)
        except (asyncio.IncompleteReadError, ConnectionError, OSError):
            pass
        finally:
            self._connected = False
            if self.writer is not None:
                try:
                    self.writer.close()
                except Exception:  # noqa: BLE001
                    pass
                self.writer = None
            self._emit("disconnect")

    def _on_message(self, msg: dict) -> None:
        req_id = msg.get("reqId")
        if req_id is not None and req_id in self._pending:
            fut = self._pending.pop(req_id)
            if msg.get("ev") == "error":
                fut.set_exception(PtyHostError(msg.get("message") or "host error"))
            else:
                fut.set_result(msg)
            if msg.get("ev") == "attached" and msg.get("replay"):
                replay = base64.b64decode(msg["replay"])
                if replay:
                    # A distinct "replay" event (not "output"): consumers
                    # must treat it as a full-state snapshot — the frontend
                    # wipes and replays instead of appending to content that
                    # is already rendered (which doubled TUI output after a
                    # pty-host reconnect).
                    self._emit("replay", msg.get("id"), replay)
            return
        ev = msg.get("ev")
        if ev == "output":
            data = base64.b64decode(msg.get("data") or "")
            self._emit("output", msg.get("id"), data)
        elif ev == "dropped":
            # pty-host's send buffer overflowed for this server connection —
            # output was NOT silently lost: the host dropped this connection
            # and asks us to resync so the next chunk is a full snapshot.
            self._emit("dropped", msg.get("id"))
        elif ev == "exit":
            self._emit("exit", msg.get("id"), msg.get("code"), msg.get("signal"))
        elif ev == "hello":
            self.server_version = msg.get("version")

    # --- request/response ----------------------------------------------------
    async def _request(self, op: str, payload: dict | None = None,
                       timeout_s: float = 5.0) -> dict:
        if not self._connected:
            await self.connect()
        assert self.writer is not None
        self._req_id += 1
        req_id = self._req_id
        msg = {"op": op, "reqId": req_id, **(payload or {})}
        self.writer.write((json.dumps(msg) + "\n").encode("utf-8"))
        await self.writer.drain()
        fut: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[req_id] = fut
        try:
            return await asyncio.wait_for(fut, timeout=timeout_s)
        except asyncio.TimeoutError:
            self._pending.pop(req_id, None)
            raise PtyHostError(f"host {op} timed out")

    def _send(self, op: str, payload: dict | None = None) -> bool:
        if not self._connected or self.writer is None:
            return False
        try:
            self.writer.write((json.dumps({"op": op, **(payload or {})}) + "\n").encode("utf-8"))
            return True
        except Exception as err:  # noqa: BLE001
            log_error("pty-host-client", err)
            return False

    # --- operations -----------------------------------------------------------
    async def list(self) -> dict:
        return await self._request("list")

    async def start(self, opts: dict) -> dict:
        return await self._request("start", opts)

    async def attach(self, sid: str) -> dict:
        return await self._request("attach", {"id": sid})

    def detach(self, sid: str) -> bool:
        return self._send("detach", {"id": sid})

    async def kill(self, sid: str) -> dict:
        return await self._request("kill", {"id": sid})

    async def forget(self, sid: str) -> dict:
        return await self._request("forget", {"id": sid})

    def input(self, sid: str, data) -> bool:  # type: ignore[no-untyped-def]
        if isinstance(data, bytes):
            raw = data
        else:
            raw = str(data).encode("utf-8")
        return self._send("input", {"id": sid, "data": base64.b64encode(raw).decode("ascii")})

    def resize(self, sid: str, cols: int, rows: int) -> bool:
        return self._send("resize", {"id": sid, "cols": cols, "rows": rows})


# --- host spawn helpers -------------------------------------------------------
async def _try_connect(timeout_s: float = 0.6) -> None:
    async def attempt() -> None:
        reader, writer = await asyncio.open_unix_connection(PIPE_NAME)
        writer.close()
        await writer.wait_closed()

    await asyncio.wait_for(attempt(), timeout=timeout_s)


async def _ensure_host_running() -> None:
    try:
        await _try_connect()
        return
    except (OSError, asyncio.TimeoutError):
        pass
    if os.name != "posix":
        raise PtyHostError(
            "pty-host requires POSIX — on Windows install pywinpty and run "
            "python src/pty_host_windows.py "
            "(pip install -r requirements-windows.txt)"
        )
    # Not up — spawn detached. Do NOT poll 4s here (audit F4: it blocked
    # server startup before the listener even bound); the host monitor in
    # SessionManager retries connect + reattaches, so fail fast and let it.
    import subprocess

    devnull = os.open(os.devnull, os.O_RDWR)
    try:
        subprocess.Popen(
            [sys.executable, HOST_SCRIPT],
            stdin=devnull, stdout=devnull, stderr=devnull,
            start_new_session=True, close_fds=True,
        )
    finally:
        os.close(devnull)
    # One short connect attempt (~300ms) to catch immediate spawn failures,
    # then hand off to the monitor.
    for _ in range(3):
        await asyncio.sleep(0.1)
        try:
            await _try_connect(0.3)
            return
        except (OSError, asyncio.TimeoutError):
            continue
    raise PtyHostError(f"pty-host did not come up at {PIPE_NAME}")

File: src/pty-host-client.js
// PTY host client. webpty connects to a separately-running pty-host daemon
// over a named pipe / Unix socket, so PTYs (claude, codex, pwsh, …) survive
// webpty restarts. If the daemon isn't running we detach-spawn it.

import net from 'node:net';
import os from 'node:os';
import path from 'node:path';
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import { EventEmitter } from 'node:events';

const PIPE_NAME = process.env.WEBPTY_PTY_HOST_PIPE
  || (process.platform === 'win32'
    ? '\\\\.\\pipe\\webpty-pty-host'
    : path.join(os.tmpdir(), 'webpty-pty-host.sock'));

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const HOST_SCRIPT = path.join(__dirname, '..', 'bin', 'pty-host.mjs');

export class PtyHostClient extends EventEmitter {
  constructor() {
    super();
    this.socket = null;
    this.buf = '';
    this.reqId = 0;
    this.pending = new Map(); // reqId → {resolve, reject}
    this.serverVersion = null;
  }

  async connect() {
    if (this.socket) return;
    if (!this._connectPromise) {
      this._connectPromise = (async () => {
        await ensureHostRunning();
        await this.#openSocket();
      })().finally(() => { this._connectPromise = null; });
    }
    await this._connectPromise;
  }

  async #openSocket() {
    await new Promise((resolve, reject) => {
      const s = net.connect(PIPE_NAME);
      const onErr = (err) => { s.removeAllListeners(); reject(err); };
      s.once('error', onErr);
      s.once('connect', () => {
        s.removeListener('error', onErr);
        this.socket = s;
        s.on('data', (chunk) => this.#onData(chunk));
        s.on('close', () => {
          this.socket = null;
          this.emit('disconnect');
        });
        s.on('error', (err) => this.emit('socket-error', err));
        resolve();
      });
    });
  }

  #onData(chunk) {
    this.buf += chunk.toString('utf8');
    let nl;
    while ((nl = this.buf.indexOf('\n')) >= 0) {
      const line = this.buf.slice(0, nl);
      this.buf = this.buf.slice(nl + 1);
      if (!line.trim()) continue;
      let msg;
      try { msg = JSON.parse(line); } catch { continue; }
      this.#onMessage(msg);
    }
  }

  #onMessage(msg) {
    // Reply to a request? Resolve the matching promise first.
    if (msg.reqId && this.pending.has(msg.reqId)) {
      const { resolve, reject } = this.pending.get(msg.reqId);
      this.pending.delete(msg.reqId);
      if (msg.ev === 'error') reject(new Error(msg.message || 'host error'));
      else resolve(msg);
      // Don't return — 'attached' replies also include replay bytes we may
      // want to surface as a synthetic 'output' so listeners see them.
      if (msg.ev === 'attached' && msg.replay) {
        const buf = Buffer.from(msg.replay, 'base64');
        if (buf.length) this.emit('output', msg.id, buf);
      }
      return;
    }
    // Spontaneous events.
    if (msg.ev === 'output') {
      this.emit('output', msg.id, Buffer.from(msg.data, 'base64'));
    } else if (msg.ev === 'exit') {
      this.emit('exit', msg.id, msg.code, msg.signal);
    } else if (msg.ev === 'hello') {
      this.serverVersion = msg.version;
      this.emit('hello', msg.version, msg.pid);
    }
  }

  async #request(op, payload = {}, timeoutMs = 5000) {
    if (!this.socket) await this.connect();
    const sock = this.socket;
    if (!sock) throw new Error('not connected');
    const reqId = ++this.reqId;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        if (!this.pending.has(reqId)) return;
        this.pending.delete(reqId);
        reject(new Error(`host ${op} timed out`));
      }, timeoutMs);
      this.pending.set(reqId, {
        resolve: (v) => { clearTimeout(timer); resolve(v); },
        reject: (e) => { clearTimeout(timer); reject(e); }
      });
      try {
        sock.write(JSON.stringify({ op, reqId, ...payload }) + '\n');
      } catch (err) {
        clearTimeout(timer);
        this.pending.delete(reqId);
        reject(err);
      }
    });
  }

  #send(op, payload = {}) {
    if (!this.socket) return false;
    try {
      this.socket.write(JSON.stringify({ op, ...payload }) + '\n');
      return true;
    } catch { return false; }
  }

  list() { return this.#request('list'); }
  start(opts) { return this.#request('start', opts); }
  attach(id) { return this.#request('attach', { id }); }
  detach(id) { return this.#send('detach', { id }); }
  kill(id) { return this.#request('kill', { id }); }
  forget(id) { return this.#request('forget', { id }); }
  input(id, data) {
    const b64 = Buffer.isBuffer(data)
      ? data.toString('base64')
      : Buffer.from(String(data), 'utf8').toString('base64');
    return this.#send('input', { id, data: b64 });
  }
  resize(id, cols, rows) { return this.#send('resize', { id, cols, rows }); }
}

// --- helpers ---------------------------------------------------------------
async function tryConnect(timeoutMs = 600) {
  return new Promise((resolve, reject) => {
    const s = net.connect(PIPE_NAME);
    const timer = setTimeout(() => {
      s.destroy();
      reject(new Error('connect timeout'));
    }, timeoutMs);
    s.once('connect', () => {
      clearTimeout(timer);
      s.end();
      resolve();
    });
    s.once('error', (err) => {
      clearTimeout(timer);
      reject(err);
    });
  });
}

async function ensureHostRunning() {
  try { await tryConnect(); return; } catch {}
  // Not up — spawn detached, then poll.
  spawnHost();
  for (let i = 0; i < 40; i++) {
    await sleep(100);
    try { await tryConnect(300); return; } catch {}
  }
  throw new Error(`pty-host did not come up at ${PIPE_NAME}`);
}

function spawnHost() {
  const child = spawn(process.execPath, [HOST_SCRIPT], {
    detached: true,
    stdio: 'ignore',
    windowsHide: true
  });
  child.unref();
}

function sleep(ms) { return new Promise((r) => setTimeout(r, ms)); }

export const PTY_HOST_PIPE_NAME = PIPE_NAME;
